### ISSUE TITLE
cctest/test-thread-termination/TerminateMultipleV8ThreadsDefaultIsolate timeout

### ISSUE THREAD

**Issue by kilvadyb** [v8mips/v8mips#36]
This test fails with timeout about 50% of the time.

**Comment by kilvadyb** [comment]
I couldn't reproduce the problem yet but from timings it can be seen that if the hjdev machine under heavy load (jenkins - d8 tests) then the test-thread-termination/TerminateMultipleV8ThreadsDefaultIsolate test slows down most:

test-thread-termination/TerminateOnlyV8ThreadFromOtherThread
real    0m0.020s
user    0m0.020s
sys 0m0.000s

test-thread-termination/TerminateMultipleV8ThreadsDefaultIsolate
real    0m47.418s
user    0m47.440s
sys 0m0.040s

test-thread-termination/TerminateLoadICException
real    0m0.013s
user    0m0.020s
sys 0m0.000s

test-thread-termination/TerminateAndReenterFromThreadItself
real    0m0.020s
user    0m0.020s
sys 0m0.000s

The make mipsel.release.check sets the timeout to 120 sec. by default (47 sec, the slowest I could reach for the critical test case is still good enough).

The test runs 2 threads with a quite simple body:
   v8::Handlev8::String source =
        v8::String::New("try { loop(); fail(); } catch(e) { fail(); }");
    v8::Script::Compile(source)->Run();
    context.Dispose(context->GetIsolate());

So it doesn't seem to be a MIPS problem for me expect if our simulator is very slow for some reason.

**Comment by paul99** [comment]
Running just the one test, in 3 modes, I see crazy long execution time:

```
$ time tools/run-tests.py -j8 --arch=mipsel --mode=release --timeout=240 --time --no-presubmit --report cctest/test-thread-termination/TerminateMultipleV8ThreadsDefaultIsolate 
>>> Running tests for mipsel.release
Total: 1 tests
 *    0 tests will be skipped
 *    0 tests are expected to timeout sometimes
 *    0 tests are expected to be flaky but not crash
 *    1 tests are expected to pass
 *    0 tests are expected to fail that we won't fix
 *    0 tests are expected to fail that we should fix
No connection to distribution server; running tests locally.
=== cctest/test-thread-termination/TerminateMultipleV8ThreadsDefaultIsolate ===
Command: /opt/paul/v8/out/mipsel.release/cctest test-thread-termination/TerminateMultipleV8ThreadsDefaultIsolate --nobreak-on-abort --nodead-code-elimination --nofold-constants --testing_serialization_file=/opt/paul/v8/out/.serdes/serdes_TerminateMultipleV8ThreadsDefaultIsolate
--- TIMEOUT ---
=== cctest/test-thread-termination/TerminateMultipleV8ThreadsDefaultIsolate ===
Command: /opt/paul/v8/out/mipsel.release/cctest --stress-opt --always-opt test-thread-termination/TerminateMultipleV8ThreadsDefaultIsolate --nobreak-on-abort --nodead-code-elimination --nofold-constants --testing_serialization_file=/opt/paul/v8/out/.serdes/serdes_TerminateMultipleV8ThreadsDefaultIsolate__stress_opt__always_opt
--- TIMEOUT ---
[16:00|% 100|+   1|-   2]: Done                                               

--- Total time: 16:00.099 ---
   1 (16:00.092) cctest/test-thread-termination/TerminateMultipleV8ThreadsDefaultIsolate
   2 (04:00.084) cctest/test-thread-termination/TerminateMultipleV8ThreadsDefaultIsolate
   3 (02:40.691) cctest/test-thread-termination/TerminateMultipleV8ThreadsDefaultIsolate

real    16m0.265s
user    22m43.949s
sys 0m1.772s
```

This is on google bleeding_edge, at this rev:

```
commit 5b41456cafa25c8a90ca3ae95d4916e57cec8ad4
Author: jkummerow@chromium.org <jkummerow@chromium.org@ce2b1a6d-e550-0410-aec6-3dcde31c8c00>
Date:   Tue Apr 16 15:17:27 2013 +0000
```

It is on a fast buildmachine, otherwise idle, run with -j8

**Comment by paul99** [comment]
The machine in previous test is a 12-core Intel(R) Xeon(R) CPU E5-2667, running Ubuntu 12.04, gcc (Ubuntu/Linaro 4.6.3-1ubuntu5) 4.6.3, Python 2.7.3.

Results below are from a Intel(R) Core(TM) i7 980x, running 10.04, gcc (Ubuntu 4.4.3-4ubuntu5.1) 4.4.3, Python 2.6.5.

This is 4 runs of the single test, with the usual 3 variations. Note the run-run timing variability, and in the 3rd run, the second test took over 3 min also.

```
$ time tools/run-tests.py  --arch=mipsel --mode=release --timeout=480 --time --no-presubmit --report cctest/test-thread-termination/TerminateMultipleV8ThreadsDefaultIsolate


[03:06|% 100|+   3|-   0]: Done                                                

--- Total time: 03:06.718 ---
   1 (03:06.707) cctest/test-thread-termination/TerminateMultipleV8ThreadsDefaultIsolate
   2 (00:53.966) cctest/test-thread-termination/TerminateMultipleV8ThreadsDefaultIsolate
   3 (00:34.843) cctest/test-thread-termination/TerminateMultipleV8ThreadsDefaultIsolate

[05:39|% 100|+   3|-   0]: Done                                                

--- Total time: 05:39.690 ---
   1 (05:39.680) cctest/test-thread-termination/TerminateMultipleV8ThreadsDefaultIsolate
   2 (00:36.141) cctest/test-thread-termination/TerminateMultipleV8ThreadsDefaultIsolate
   3 (00:21.929) cctest/test-thread-termination/TerminateMultipleV8ThreadsDefaultIsolate

[03:21|% 100|+   3|-   0]: Done                                                

--- Total time: 03:21.637 ---
   1 (03:21.627) cctest/test-thread-termination/TerminateMultipleV8ThreadsDefaultIsolate
   2 (03:07.811) cctest/test-thread-termination/TerminateMultipleV8ThreadsDefaultIsolate
   3 (00:19.528) cctest/test-thread-termination/TerminateMultipleV8ThreadsDefaultIsolate

[02:32|% 100|+   3|-   0]: Done                                                

--- Total time: 02:32.487 ---
   1 (02:32.477) cctest/test-thread-termination/TerminateMultipleV8ThreadsDefaultIsolate
   2 (01:11.373) cctest/test-thread-termination/TerminateMultipleV8ThreadsDefaultIsolate
   3 (00:39.048) cctest/test-thread-termination/TerminateMultipleV8ThreadsDefaultIsolate
```

**Comment by kilvadyb** [comment]
I managed to catch timeouts when the strong machine was otherwise idle. The test starts 2 threads which signals back to the starter and it terminates the two threads. The 2 test threads run in an infinite loop, sending signals in the first step of the loop. The Run() function of the Looping threads acquires a lock before executing the javascript loop:
  void Run() {
    v8::Locker locker(CcTest::default_isolate());

When the test stuck in, With gdb I saw that one of the threads (Thread 7) running in the loop (of its javascript code) while the other (Thread 8) is waiting for the initial mutex (Locker):
Thread 7 (Thread 0xf73d0b70 (LWP 23077)):
#0  0x08512b82 in v8::internal::Simulator::DecodeTypeRegister(v8::internal::Instruction_) ()
#1  0x085134fc in v8::internal::Simulator::InstructionDecode(v8::internal::Instruction_) ()
#2  0x08515303 in v8::internal::Simulator::CallInternal(unsigned char_) ()
#3  0x085153da in v8::internal::Simulator::Call(unsigned char_, int, ...) ()
#4  0x08259c22 in v8::internal::Invoke(bool, v8::internal::Handlev8::internal::JSFunction, v8::internal::Handlev8::internal::Object, int, v8::internal::Handlev8::internal::Object_, bool_) ()
#5  0x0825a918 in v8::internal::Execution::Call(v8::internal::Handlev8::internal::Object, v8::internal::Handlev8::internal::Object, int, v8::internal::Handlev8::internal::Object_, bool_, bool) ()
#6  0x081f9756 in v8::Script::Run() ()
#7  0x081c5ec8 in DoLoop(v8::Arguments const&) ()

Thread 8 (Thread 0xf6bcfb70 (LWP 23078)):
#1  0xf7e74ed9 in __lll_lock_wait () at ../nptl/sysdeps/unix/sysv/linux/i386/i686/../i486/lowlevellock.S:142
#2  0xf7e70219 in _L_lock_835 () from /lib32/libpthread.so.0
#3  0xf7e700ab in __pthread_mutex_lock (mutex=0x8f9b854) at pthread_mutex_lock.c:82
#4  0x0846a442 in v8::Locker::Initialize(v8::Isolate_) ()
#5  0x081c5340 in LoopingThread::Run() ()
#6  0x08525d66 in v8::internal::ThreadEntry(void_) ()

Thread 7 will release the Lock/mutex when it is terminated but the starter main thread terminates both of the test threads when it receives both of the signals:
  for (int i = 0; i < kThreads; i++) {
    semaphore->Wait();
  }
  {
    v8::Locker locker(CcTest::default_isolate());
    for (int i = 0; i < kThreads; i++) {
      v8::V8::TerminateExecution(threads[i]->GetV8ThreadId());
    }
  }

Documentions says:
 /**
- Multiple threads in V8 are allowed, but only one thread at a time is allowed
- to use any given V8 isolate

I have to do further testing to find when the test works and when it sticks in.

**Comment by kilvadyb** [comment]
I did a lot of work/try to debug the timeout problem on otherwise idle mac and linux machine with 4 cores (2 real cores with hyper threading) without luck. I could reproduce it on hjdev machine when only this cctest case is running with release and debug executables. I made coredumps from gdb to make it possible to examine the situation later. There might be some relationship between free cores and number of running/existing threads (8 thread exists in this cctest case: 4 sweeper, 1 context swather, 2 test threads and the main starter thread).

The starter and the 2 test threads are sharing the same default isolate instance and only one thread can use an isolate at a time. The isolate usage is saved with Locker (pthread_mutex) and a thread can yield other by unlocking it (context switcher sets a flag of the isolate for PREEMPT and the context switch happens when a running test thread execution in the simulator reaches a software interrupt.

One of the test threads (A) can stick in when the other thread (B) already reached the javascript execution (simulator code is running) and (A) is trying to initialize its Locker (lock the common isolate's pthread_mutex). When both thread executes its javascript code then context switching seems to work fine (RuntimePreempt function finds the other thread and yields it). The starter thread also can stick in (hang on the isolate's mutex) because it must initialize its own Locker to the isolate before it can wait for the semaphore signals of the test threads. If one or both of the test threads are executing javascript (simulator active ?) when the starter tries to acquire the lock/mutex then the starter thread might be blocked by the mutex for long time.

I have to further analyze the coredumps and now I have this idea:

When context switching happens (on V8 thread level) the isolate's mutex gets unlocked and locked back so other (linux level) threads might be activated. It seems when an 8 core machine runs only this test then pthread_mutex_unlock rarely activates threads outside the simulator (other core?). If the hjdev machine has other heavy loads then the linux level context switching happens frequently and pthread_mutex_unlock works "fine". pthread_mutex_unlock uses glib's implementation which context switches to kernel (via futex system call) only if there is no other way. It's all just theory as google is not so close friend of mine - I mean I haven't found reliable documentations about it yet.

If any of you knows anything about pthread_mutexes, pthreads on hardware cores then don't hesitate to enlighten me, please.

**Comment by kilvadyb** [comment]
I made the same test with arm simulator running the following script as for the mips-sim:
1 2 3 4 5 6 7 8 9; do for i in 1 2 3 4 5 6 7 8 9; do echo -e "\n$j:$i"; time out/arm.release/cctest --stress-opt --always-opt test-thread-termination/TerminateMultipleV8ThreadsDefaultIsolate --nobreak-on-abort --nodead-code-elimination --nofold-constants; done; done 2>&1 | tee log-arm.txt

I think respense time/scheduling could be improved with more context switching from simulator (currently evry 3rd SoftwerInterrupt calls a V8::Thread::Yield() == sched_yield()). But it might slow down the simulator running other tests.

The results of both platform-simulator tests are very similar on my i5 dual core (4 hyperthreaded cores). When the machine is otherwise idle the results varies between 2 minutes and a few seconds, when there is other load then times are between 0.0xs and 0.3xs:
selected results for MIPS:
1:1
real    1m4.561s

1:2
real    0m13.470s

1:3
real    0m16.126s

1:4
real    0m12.386s

3:5
real    2m13.504s

3:6
real    0m24.936s

When I started for example chromium which caused a lot of low level context switching (core usage?):
2:4
real    0m0.040s

2:5
real    0m0.344s

2:6
real    0m0.378s

On arm:
When machine is otherwise idle:
5:5
real    1m50.161s

5:6
real    0m21.873s

5:7
real    0m42.460s

5:8
real    0m15.943s

With other load:
8:7
real    0m0.029s

8:8
real    0m0.040s

**Comment by kilvadyb** [comment]
Running the test on arm-sim on the otherwise idle hjdev (8 core machine) I saw similar timeouts sometimes, like on mips-sim:

for i in 1 2 3 4 5 6 7 8 9 10; do echo -e "\n\n$i:"; time out/arm.release/cctest --stress-opt --always-opt test-thread-termination/TerminateMultipleV8ThreadsDefaultIsolate --nobreak-on-abort --nodead-code-elimination --nofold-constants; done

One of the results:
2:

real    7m35.508s
user    7m36.020s
sys 0m0.900s

On arm there are similar stack backtraces of threads when they stick in (thread 1 hangs on the isolate's mutex, thread 2 and 3 are working context switching between themselves / both are executing javascript DoLoop):

Thread 1 (Thread 0xf74f7b30 (LWP 23989)):
#0  0xf77cf430 in __kernel_vsyscall ()
#1  0xf7664ed9 in __lll_lock_wait () at ../nptl/sysdeps/unix/sysv/linux/i386/i686/../i486/lowlevellock.S:142
#2  0xf7660219 in _L_lock_835 () from /lib32/libpthread.so.0
#3  0xf76600ab in __pthread_mutex_lock (mutex=0x9d72854) at pthread_mutex_lock.c:82
#4  0x084707a2 in v8::Locker::Initialize(v8::Isolate*) ()
#5  0x081c4f01 in TestTerminateMultipleV8ThreadsDefaultIsolate() ()
#6  0x0804b26e in main ()

Thread 2 (Thread 0xf63bfb70 (LWP 23996)):
#0  0x085214f6 in v8::internal::Simulator::GetShiftRm(v8::internal::Instruction_, bool_) ()
#1  0x08526b36 in v8::internal::Simulator::DecodeType01(v8::internal::Instruction_) ()
#2  0x08528014 in v8::internal::Simulator::InstructionDecode(v8::internal::Instruction_) ()
#3  0x085284d3 in v8::internal::Simulator::CallInternal(unsigned char_) ()
#4  0x085285cb in v8::internal::Simulator::Call(unsigned char_, int, ...) ()
#5  0x0825e322 in v8::internal::Invoke(bool, v8::internal::Handlev8::internal::JSFunction, v8::internal::Handlev8::internal::Object, int, v8::internal::Handlev8::internal::Object_, bool_) ()
#6  0x0825f018 in v8::internal::Execution::Call(v8::internal::Handlev8::internal::Object, v8::internal::Handlev8::internal::Object, int, v8::internal::Handlev8::internal::Object_, bool_, bool) ()
#7  0x081fd596 in v8::Script::Run() ()
#8  0x081c5b48 in DoLoop(v8::Arguments const&) ()
#9  0x08226446 in v8::internal::Builtin_HandleApiCall(int, v8::internal::Object*_, v8::internal::Isolate_) ()
#10 0x085276f2 in v8::internal::Simulator::SoftwareInterrupt(v8::internal::Instruction_) ()
#11 0x0852802c in v8::internal::Simulator::InstructionDecode(v8::internal::Instruction_) ()
#12 0x085284d3 in v8::internal::Simulator::CallInternal(unsigned char_) ()
#13 0x085285cb in v8::internal::Simulator::Call(unsigned char_, int, ...) ()
#14 0x0825e322 in v8::internal::Invoke(bool, v8::internal::Handlev8::internal::JSFunction, v8::internal::Handlev8::internal::Object, int, v8::internal::Handlev8::internal::Object_, bool_) ()
#15 0x0825f018 in v8::internal::Execution::Call(v8::internal::Handlev8::internal::Object, v8::internal::Handlev8::internal::Object, int, v8::internal::Handlev8::internal::Object_, bool_, bool) ()
(More stack frames follow...)

Thread 3 (Thread 0xf6bc0b70 (LWP 23995)):
#0  0xf77cf430 in __kernel_vsyscall ()
#1  0xf7664ed9 in __lll_lock_wait () at ../nptl/sysdeps/unix/sysv/linux/i386/i686/../i486/lowlevellock.S:142
#2  0xf7660219 in _L_lock_835 () from /lib32/libpthread.so.0
#3  0xf76600ab in __pthread_mutex_lock (mutex=0x9d72854) at pthread_mutex_lock.c:82
#4  0x084706a1 in v8::Unlocker::~Unlocker() ()
#5  0x082608aa in v8::internal::Execution::HandleStackGuardInterrupt(v8::internal::Isolate_) ()
#6  0x085276f2 in v8::internal::Simulator::SoftwareInterrupt(v8::internal::Instruction_) ()
#7  0x0852802c in v8::internal::Simulator::InstructionDecode(v8::internal::Instruction_) ()
#8  0x085284d3 in v8::internal::Simulator::CallInternal(unsigned char_) ()
#9  0x085285cb in v8::internal::Simulator::Call(unsigned char_, int, ...) ()
#10 0x0825e322 in v8::internal::Invoke(bool, v8::internal::Handlev8::internal::JSFunction, v8::internal::Handlev8::internal::Object, int, v8::internal::Handlev8::internal::Object_, bool_) ()
#11 0x0825f018 in v8::internal::Execution::Call(v8::internal::Handlev8::internal::Object, v8::internal::Handlev8::internal::Object, int, v8::internal::Handlev8::internal::Object_, bool_, bool) ()
#12 0x081fd596 in v8::Script::Run() ()
#13 0x081c5b48 in DoLoop(v8::Arguments const&) ()
#14 0x08226446 in v8::internal::Builtin_HandleApiCall(int, v8::internal::Object__, v8::internal::Isolate_) ()
#15 0x085276f2 in v8::internal::Simulator::SoftwareInterrupt(v8::internal::Instruction*) ()

**Comment by paul99** [comment]
> I think respense time/scheduling could be improved with more context switching from simulator 
> (currently evry 3rd SoftwareInterrupt calls a V8::Thread::Yield() == sched_yield()). But it might 
> slow down the simulator running other tests.

I don't see where this Yield() is called in the simulator. Can you be more specific about how this happens?

If I understood you correctly, you are saying that unless we call out to runtime, the simulator will never give up the cpu. We know our simulator is slow. Maybe we can find a place (at least for some testing) to do the cpu yield more often within the simulator execution loop.

You've said that ARM has similar issues. Presumably if this is a simulator time-hog issue, then ia32 & x64 would never have this problem. Is that worth testing?

**Comment by kilvadyb** [comment]
> I don't see where this Yield() is called in the simulator. Can you be more specific about how this happens?

This test uses one default isolate instance to run 2 test threads and one main or starter thread. As docs says "only one thread at a time is allowed to use any given V8 isolate". Isolate sharing happens the following way:
A thread must lock the used isolate, for example test threads of this cctest runs this thread code:

```
class LoopingThread : public v8::internal::Thread {
 public:
  LoopingThread() : Thread("LoopingThread") { }
  void Run() {
    v8::Locker locker(CcTest::default_isolate()); // Lock the isolate's mutex
    …
    // execute javascript endless loop.
  } // locker's dtor unlocks the isolate's mutex
```

To let threads run parallel using the same isolate a ContextSwitcher thread is running:

```
// The ContextSwitcher thread is used to schedule regular preemptions to
// multiple running V8 threads.

src/v8threads.cc:
// Main loop of the ContextSwitcher thread: Preempt the currently running V8
// thread at regular intervals.
void ContextSwitcher::Run() {
  while (keep_going_) {
    OS::Sleep(sleep_ms_);
    isolate()->stack_guard()->Preempt();
  }
}
```

`isolate()->stack_guard()->Preempt();` call sets a flag for preemption/yielding:
`thread_local_.interrupt_flags_ |= PREEMPT;`

When simulator (executing javascript code of the current thread) reaches a Software Interrupt and runs a HandleStackGuardInterrupt from it and the PREEMPT flag of the thread is set/high then a 'forced' context switching happens. Stacktrace of it:

#3 0xf76600ab in __pthread_mutex_lock (mutex=0x9d72854) at pthread_mutex_lock.c:82
#4 0x084706a1 in v8::Unlocker::~Unlocker() ()
#5 0x082608aa in v8::internal::Execution::HandleStackGuardInterrupt(v8::internal::Isolate_) ()
#6 0x085276f2 in v8::internal::Simulator::SoftwareInterrupt(v8::internal::Instruction_) ()
#7 0x0852802c in v8::internal::Simulator::InstructionDecode(v8::internal::Instruction*) ()

The Unlocker class is a guard similar to the Locker but the ctor releases/unlocks the isolate's mutex while the dtor  acquires/locks it again so at the calling the ctor of it an other thread -- pending on the isolate object's mutex -- can be wakened.

The code of the context switching (happens from the context of a V8 thread) where the Unlocker dtor is reached in the above stack trace:

```
// src/execution.cc:886
MaybeObject* Execution::HandleStackGuardInterrupt(Isolate* isolate) {
…
  if (stack_guard->IsPreempted()) RuntimePreempt(); // if PREEMPT flag of current thread is set/high
…
}

// src/execution.cc:775
static Object* RuntimePreempt() {
…
  } else {
    // Perform preemption.
    v8::Unlocker unlocker(reinterpret_cast<v8::Isolate*>(isolate)); // unlock the isolate's mutex.
    Thread::YieldCPU(); // call int sched_yield(void); sys call.
  } // door of unlocker locks the isolate's mutex.
…
}
```

So YieldCPU() is here. A context switching happens from simulator's SoftwerInterrupt handling so if we want to modify only MIPS related code then we can try to make SoftwerInterrupts or v8::internal::Execution::HandleStackGuardInterrupt calls from SoftwerInterrupt/instruction decoding loop be more frequent.

> If I understood you correctly, you are saying that unless we call out to runtime, the simulator will never give up the cpu. We know our simulator is slow. Maybe we can find a place (at least for some testing) to do the cpum yield more often within the simulator execution loop.

As I see the code that's what we can try but I think at first the linux mutex handling should be investigated without simulator (i.e. testing x64 native run).

> You've said that ARM has similar issues. Presumably if this is a simulator time-hog issue, then ia32 & x64 would never have this problem. Is that worth testing?

It's not just the simulator. I guess it is related with free CPU cores/pthread_mutex_unlock handling (kernel context switching) as the problem didn't come up with heavy load or on mac. I tested on the same mac machine and linux produced much longer waits when the machine is otherwise idle. If there is some load (like I started a chromium browser which caused a lot of kernel context switching, maybe changes in core usage) then a long time pending test finished quickly and new tests (from a bash loop) run very quickly on MIPS or ARM simulator. On mac the pthread_mutexes are implemented with the Mach microkernel's features not in the unix kernel. But I should make some tests also when no other programs (Skype, mail, ….) are running. I haven't found reliable description about linux mutexes but one of the sources said that on linux glib (and kernel futex) has a pthread_mutex implementation that at unlocking wakes up a waiting thread on the same CPU core as it is quicker then doing a CPU core switch. Other source said it happens only with default mutexes but V8 is using recursive mutexes (`pthread_mutexattr_settype(&attrs, PTHREAD_MUTEX_RECURSIVE);`). So I think it worths a try on x64/ia32 without simulator as the V8 thread context switching and mutex usage must be similar.

**Comment by paul99** [comment]
Please do try this on ia32 & x64, and let us know what happens. Be also great to know if it does show up on Linux, does it not happen on Mac, as with the mips version.

If the problem shows in ia32/x64, it is possible that we could create a V8 bug, explain what is seen, and possibly get some help in debugging this.

It is also interesting to me that only this test has the issue, and there are many more threading tests. Can we see what might be 'unique' about this one?

I will look more at the other information you provided tomorrow, thanks!

**Comment by kisg** [comment]
Results from Friday, forgot to copy in here: 
We checked it on mac and ubuntu linux x64, on the same mac hardware. On mac there was no slowdown, even when in idle. On Ubuntu:
run 19
real    0m39.501s
user    0m39.680s
sys     0m0.176s

run 20
real    0m11.172s
user    0m11.184s
sys     0m0.072s

run 21
real    0m0.033s
user    0m0.028s
sys     0m0.004s

As you can see, while not as severe as for the simulators, it can still get into a locked state. During run 21 a firefox instance was started, and the test finished almost instantly. So this looks like an issue with the the process scheduler of the linux kernel. It looks like it affects a lot of kernel versions (we tested on Ubuntu lucid - 2.6.32, and also on the latest Ubuntu 3.5 or later -- I am not sure which)

Balázs: I see 2 possibilities:
1. You could try to force the threads to different cores using pthread_setaffinity_np, and see if the problem still persists.
2. Add some extra logging to see if the threads are really always scheduled on the same core while they are locked. You would have to use a kernel tracing framework like systemtap or dtrace for this, because the userspace has no reliable information about the current cpu -- it could be rescheduled to a different core before the next instruction runs without notice. (To make the resulting logs smaller, you should probably only log, when the assigned core changes.)

**Comment by kisg** [comment]
I see now that we never included it in here: Since this happens on ARM as well, we opened an issue on the official v8 tracker:
https://code.google.com/p/v8/issues/detail?id=2657

No activity there yet.

**Comment by kilvadyb** [comment]
I made the pthread_setaffinity_np test Gergely suggested. If all the 3 threads which acquires the mutex of the default isolate are forced to run on the same core then the test terminates very quickly (time reports 0m0.0x sec runtimes). If test threads are on core 1 and the main/starter/waiter thread is on core 2 then I got similar runtimes as without forcing the cores. Starting other programs (disk usage, new process activity - core usage?) also speeds up pending tests. Nice questions: What about other threading test? (when more threads are using the same isolate then there might be similar problems) And could we reproduce the same locking/sticking problem using pure linux pthreads without any V8 or other code (this test could be implemented without javascript code).

**Comment by kilvadyb** [comment]
Diff of the pthread_setaffinity_np:

```
diff --git a/test/cctest/test-thread-termination.cc b/test/cctest/test-thread-termination.cc
index 190fc7b..14d656a 100644
--- a/test/cctest/test-thread-termination.cc
+++ b/test/cctest/test-thread-termination.cc
@@ -29,6 +29,17 @@
 #include "platform.h"
 #include "cctest.h"

+#include <pthread.h>
+
+static int _setCpuAffinity(int core_id) {
+  cpu_set_t cpuset;
+  CPU_ZERO(&cpuset);
+  CPU_SET(core_id, &cpuset);
+
+  pthread_t current_thread = pthread_self();    
+  int res = pthread_setaffinity_np(current_thread, sizeof(cpu_set_t), &cpuset);
+  return res;
+}

 v8::internal::Semaphore* semaphore = NULL;

@@ -202,6 +213,7 @@ class LoopingThread : public v8::internal::Thread {
  public:
   LoopingThread() : Thread("LoopingThread") { }
   void Run() {
+    _setCpuAffinity(1);
     v8::Locker locker(CcTest::default_isolate());
     v8::HandleScope scope(CcTest::default_isolate());
     v8_thread_id_ = v8::V8::GetCurrentThreadId();
@@ -233,6 +245,7 @@ TEST(TerminateMultipleV8ThreadsDefaultIsolate) {
     v8::Locker::StartPreemption(1);
     semaphore = v8::internal::OS::CreateSemaphore(0);
   }
+  _setCpuAffinity(1);
   const int kThreads = 2;
   i::List<LoopingThread*> threads(kThreads);
   for (int i = 0; i < kThreads; i++) {
```

**Comment by paul99** [comment]
This spurious failure is hitting us a lot on our internal buildbot.  Do you think we should disable the test (maybe mips only) until we get this fixed?

**Comment by paul99** [comment]
Does this test fail on real MIPS hardware, or only on the sim?

**Comment by paul99** [comment]
Balázs, can you please comment on the hardware vs sim question above.

I am planning to put this test on the skip list for MIPS, because it is affecting out buildbot pretty badly.

Anyone want to argue that it is a bad idea to do that?

**Comment by kisg** [comment]
Hi Paul, 

The issue is not MIPS specific (we see it on real x86 and x86 with ARM simu as well), and only affects multi-core systems. One workaround is to force the cpu affinity settings to a single core. Balázs will try reproduce the problem on our dual-core MIPS board. I would prefer forcing the test to a single core instead of skipping it on MIPS.

We also tried to reproduce the issue without V8, just 3 threads waiting on a mutex, and could not reproduce it. So the bug itself might be V8 specific, relating to how the isolates work.

Best Regards,
Gergely

**Comment by kilvadyb** [comment]
I could not reproduce the timeout failure on a Broadcom BCM97425B1 dual-core board. The slowest execution time was under 0.8 s.

**Comment by paul99** [comment]
I've put this test on skip list for mips, https://codereview.chromium.org/16203005/, r15063.

In that issue Jakob commented:

> Both ia32 and x64 builds hit this (on a multi-core workstation). Sometimes the
> test finishes in less than a second, sometimes it gets killed as timeout after
> 60 seconds. Anything between happens as well. I can confirm that generating lots
> of CPU load seems to make it finish faster.

**Comment by paul99** [comment]
This failing test was deleted from V8 on Sep 6, 2013, in https://codereview.chromium.org/23538007

Closing.
